xds/src/main/java/io/grpc/xds/XdsClusterResource.java

```diff
@@ -169,9 +174,50 @@ static CdsUpdate processCluster(Cluster cluster,
     updateBuilder.filterMetadata(
         ImmutableMap.copyOf(cluster.getMetadata().getFilterMetadataMap()));
 
+    try {
+      ImmutableMap<String, Object> parsedFilterMetadata =
+          parseClusterMetadata(cluster.getMetadata());
+      updateBuilder.parsedMetadata(parsedFilterMetadata);
+    } catch (InvalidProtocolBufferException e) {
+      throw new ResourceInvalidException("xDS filter metadata invalid.");
+    }
+
     return updateBuilder.build();
   }
 
+  private static ImmutableMap<String, Object> parseClusterMetadata(Metadata metadata)
+      throws InvalidProtocolBufferException {
+    ImmutableMap.Builder<String, Object> parsedMetadata = ImmutableMap.builder();
+
+    ClusterMetadataRegistry registry = ClusterMetadataRegistry.getInstance();
+    // Process typed_filter_metadata
+    for (Map.Entry<String, Any> entry : metadata.getTypedFilterMetadataMap().entrySet()) {
+      String key = entry.getKey();
+      Any value = entry.getValue();
+      ClusterMetadataValueParser parser = registry.findParser(value.getTypeUrl());
+      if (parser != null) {
+        Object parsedValue = parser.parse(value);
+        if (parsedValue == null) {
```

Comment:
If the parser can throw `InvalidProtocolBufferException`, why also allow it to return null instead of throwing an exception? The exception created has no information about what went wrong.